src/Sylius/Bundle/PayumBundle/Form/Type/GatewayConfigType.php

```diff
@@ -0,0 +1,101 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sylius\Bundle\PayumBundle\Form\Type;
+
+use Payum\Core\Model\GatewayConfigInterface;
+use Sylius\Bundle\ResourceBundle\Form\Type\AbstractResourceType;
+use Sylius\Component\Registry\ServiceRegistryInterface;
+use Symfony\Component\Form\Extension\Core\Type\TextType;
+use Symfony\Component\Form\FormBuilderInterface;
+use Symfony\Component\Form\FormEvent;
+use Symfony\Component\Form\FormEvents;
+use Symfony\Component\OptionsResolver\OptionsResolver;
+
+/**
+ * @author Mateusz Zalewski <mateusz.zalewski@lakion.com>
+ */
+final class GatewayConfigType extends AbstractResourceType
+{
+    /**
+     * @var ServiceRegistryInterface
+     */
+    private $gatewayConfigurationTypeRegistry;
+
+    /**
+     * {@inheritdoc}
+     *
+     * @param ServiceRegistryInterface $gatewayConfigurationTypeRegistry
+     */
+    public function __construct(
+        $dataClass,
+        $validationGroups = [],
+        ServiceRegistryInterface $gatewayConfigurationTypeRegistry
+    ) {
+        parent::__construct($dataClass, $validationGroups);
+
+        $this->gatewayConfigurationTypeRegistry = $gatewayConfigurationTypeRegistry;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function buildForm(FormBuilderInterface $builder, array $options)
+    {
+        $factory = $options['factory'];
```

Comment:
Can we move it just before that `addEventListener` call? Or just use whole `$options` there?